File: scripts/aichain_e2e.py
#!/usr/bin/env python3

import argparse
import importlib
import json
import os
import pathlib
import sys


ROOT = pathlib.Path(__file__).resolve().parent.parent


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--mode", default="self-check", choices=["self-check", "sdk-python"])
    args = parser.parse_args()

    profile = load_profile(args.mode)
    fixtures = load_fixtures()
    validate_profile(profile)
    validate_fixtures(fixtures)

    if args.mode == "self-check":
        print(
            json.dumps(
                {
                    "ok": True,
                    "mode": args.mode,
                    "profile": profile,
                    "fixtures": {
                        "textZhBytes": len(fixtures["text_zh"].encode("utf-8")),
                        "textEnBytes": len(fixtures["text_en"].encode("utf-8")),
                        "zhAudioBytes": len(fixtures["zh_audio"]),
                        "enAudioBytes": len(fixtures["en_audio"]),
                        "imageBytes": len(fixtures["image"]),
                        "imageQuestion": fixtures["image_question"],
                        "imageExpected": fixtures["image_expected"],
                    },
                },
                ensure_ascii=False,
                indent=2,
            )
        )
        return 0

    run_python_sdk(profile, fixtures)
    return 0


def load_profile(mode: str) -> dict:
    if os.environ.get("AICHAIN_PROFILE_FILE"):
        return json.loads(pathlib.Path(os.environ["AICHAIN_PROFILE_FILE"]).read_text(encoding="utf-8"))
    if os.environ.get("AICHAIN_PROFILE"):
        return json.loads(os.environ["AICHAIN_PROFILE"])
    return {
        "region": {"code": os.environ.get("AICHAIN_REGION", "cn"), "endpoint": os.environ.get("AICHAIN_ENDPOINT")},
        "integration": "python_sdk" if mode == "sdk-python" else "webapi",
        "modelId": os.environ.get("AICHAIN_MODEL_ID", "b16924f42ffe4fd895d4ba4778278bc3"),
        "capabilities": {"asr": False, "nlu": True, "tts": False, "image": False},
        "asr": {"language": None, "duplex": None, "vad": None},
        "interruption": "none",
        "audio": {"input": None, "output": None, "notes": None},
        "tts": {"chunk": None},
        "runtime": {
            "appIdEnv": "AICHAIN_APP_ID",
            "appKeyEnv": "AICHAIN_APP_KEY",
            "endpointEnv": "AICHAIN_ENDPOINT",
            "snEnv": "AICHAIN_SN",
        },
        "android": {
            "language": None,
            "ui": None,
            "minSdk": None,
            "targetSdk": None,
            "credentialStrategy": None,
            "testStrategy": None,
        },
    }


def load_fixtures() -> dict:
    return {
        "text_zh": (ROOT / "assets/text/sample_zh.txt").read_text(encoding="utf-8"),
        "text_en": (ROOT / "assets/text/sample_en.txt").read_text(encoding="utf-8"),
        "zh_audio": (ROOT / "assets/audio/zh_16k_pcm.wav").read_bytes(),
        "en_audio": (ROOT / "assets/audio/en_16k_pcm.wav").read_bytes(),
        "image": (ROOT / "assets/image/apple.png").read_bytes(),
        "image_question": (ROOT / "assets/image/apple_question_zh.txt").read_text(encoding="utf-8").strip(),
        "image_expected": (ROOT / "assets/image/apple_expected_zh.txt").read_text(encoding="utf-8").strip(),
    }


def validate_profile(profile: dict) -> None:
    missing = []
    if not profile.get("region", {}).get("code"):
        missing.append("region.code")
    if not profile.get("integration"):
        missing.append("integration")
    if not profile.get("modelId"):
        missing.append("modelId")
    for key in ("asr", "nlu", "tts", "image"):
        if not isinstance(profile.get("capabilities", {}).get(key), bool):
            missing.append(f"capabilities.{key}")
    if not profile.get("runtime", {}).get("appIdEnv"):
        missing.append("runtime.appIdEnv")
    if not profile.get("runtime", {}).get("appKeyEnv") and not profile.get("runtime", {}).get("apiKeyEnv"):
        missing.append("runtime.appKeyEnv")
    if not profile.get("runtime", {}).get("endpointEnv"):
        missing.append("runtime.endpointEnv")
    if missing:
        raise SystemExit(f"Invalid AIChainProfile. Missing: {', '.join(missing)}")


def validate_fixtures(fixtures: dict) -> None:
    failed = []
    if not fixtures["text_zh"]:
        failed.append("text_zh")
    if not fixtures["text_en"]:
        failed.append("text_en")
    if not is_wav(fixtures["zh_audio"]):
        failed.append("zh_audio")
    if not is_wav(fixtures["en_audio"]):
        failed.append("en_audio")
    if not fixtures["image"].startswith(b"\x89PNG\r\n\x1a\n"):
        failed.append("image")
    if not fixtures["image_question"]:
        failed.append("image_question")
    if not fixtures["image_expected"]:
        failed.append("image_expected")
    if failed:
        raise SystemExit(f"Invalid bundled fixtures: {', '.join(failed)}")


def is_wav(data: bytes) -> bool:
    return data[:4] == b"RIFF" and data[8:12] == b"WAVE"


def require_runtime(profile: dict) -> dict:
    endpoint_env = profile["runtime"]["endpointEnv"]
    app_id_env = profile["runtime"]["appIdEnv"]
    app_key_env = profile["runtime"].get("appKeyEnv") or profile["runtime"].get("apiKeyEnv") or "AICHAIN_APP_KEY"
    fallback_app_key_env = "AICHAIN_APP_KEY" if app_key_env == "AICHAIN_API_KEY" else "AICHAIN_API_KEY"
    endpoint = os.environ.get(endpoint_env) or resolve_endpoint(profile.get("region", {}).get("code"))
    missing = []
    if not endpoint:
        missing.append(f"{endpoint_env} or mapped AICHAIN_REGION")
    if not os.environ.get(app_id_env):
        missing.append(app_id_env)
    if not os.environ.get(app_key_env) and not os.environ.get(fallback_app_key_env):
        missing.append(f"{app_key_env} or {fallback_app_key_env}")
    if missing:
        raise SystemExit(f"Missing required environment variables: {', '.join(missing)}")
    return {
        "endpoint": endpoint,
        "app_id": os.environ[app_id_env],
        "app_key": os.environ.get(app_key_env) or os.environ.get(fallback_app_key_env),
        "user_id": os.environ.get("AICHAIN_USER_ID"),
        "sn": os.environ.get("AICHAIN_SN") or os.environ.get("AICHAIN_DEVICE_ID") or "codex-smoke-001",
    }


def resolve_endpoint(region_code: str | None) -> str | None:
    endpoints = {
        "shanghai": "wss://aichain-sh.xfyun.cn",
        "us": "wss://aichain-us.iflyoversea.com",
        "rus": "wss://aichain-rus.iflyoversea.com",
    }
    return endpoints.get(str(region_code or "cn").lower())


def run_python_sdk(profile: dict, fixtures: dict) -> None:
    runtime = require_runtime(profile)
    package_name = os.environ.get("AICHAIN_PYTHON_SDK_PACKAGE")
    if not package_name:
        raise SystemExit("Missing AICHAIN_PYTHON_SDK_PACKAGE for generic Python SDK mode.")

    sdk = importlib.import_module(package_name)
    client_cls = getattr(sdk, "AIChainClient", None) or getattr(sdk, "Client", None)
    if client_cls is None:
        raise SystemExit(f"Cannot find AIChainClient or Client in {package_name}")

    client = client_cls(
        host=runtime["endpoint"],
        app_id=runtime["app_id"],
        app_key=runtime["app_key"],
        user_id=runtime["user_id"],
        sn=runtime["sn"],
        model_id=profile["modelId"],
        profile=profile,
    )

    results = []
    if profile["capabilities"]["nlu"] or profile["capabilities"]["tts"]:
        results.append(call_first(client, ("send_text", "chat"), fixtures["text_zh"], {"text": fixtures["text_zh"], "profile": profile}))
    if profile["capabilities"]["asr"]:
        audio = fixtures["en_audio"] if profile.get("asr", {}).get("language") == "en" else fixtures["zh_audio"]
        results.append(call_first(client, ("send_audio", "recognize"), audio, {"audio": audio, "profile": profile}))
    if profile["capabilities"]["image"]:
        results.append(
            call_first(
                client,
                ("send_image", "chat"),
                fixtures["image"],
                {
                    "image": fixtures["image"],
                    "prompt": fixtures["image_question"],
                    "mime_type": "image/png",
                    "profile": profile,
                },
            )
        )

    output = json.dumps(results, ensure_ascii=False, indent=2, default=str)
    print(output)
    validate_service_output(profile, output, fixtures)


def call_first(client, names, positional, object_arg):
    for name in names:
        method = getattr(client, name, None)
        if callable(method):
            try:
                return method(object_arg)
            except TypeError:
                return method(positional)
    raise SystemExit(f"SDK client must expose one of: {', '.join(names)}")


def validate_service_output(profile: dict, output: str, fixtures: dict) -> None:
    lower = output.lower()
    failures = []
    if profile["capabilities"]["asr"]:
        expected = ("hello", "weather") if profile.get("asr", {}).get("language") == "en" else ("北京", "天气")
        for term in expected:
            if term.lower() not in lower:
                failures.append(f"ASR output missing {term}")
    if profile["capabilities"]["nlu"] and not output.strip():
        failures.append("NLU output is empty")
    if profile["capabilities"]["tts"] and not any(term in lower for term in ("audio", "tts", "pcm", "opus", "wav", "base64")):
        failures.append("TTS output did not mention or contain audio")
    if profile["capabilities"]["image"] and fixtures["image_expected"].lower() not in lower:
        failures.append(f"Image output missing {fixtures['image_expected']}")
    if failures:
        raise SystemExit("AIChain E2E validation failed:\n- " + "\n- ".join(failures))


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except KeyboardInterrupt:
        raise SystemExit(130)

File: scripts/aichain-e2e.mjs
#!/usr/bin/env node

import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const args = new Map();
for (let i = 2; i < process.argv.length; i += 1) {
  const item = process.argv[i];
  if (item.startsWith("--")) {
    const [key, inlineValue] = item.slice(2).split("=", 2);
    const next = process.argv[i + 1];
    if (inlineValue !== undefined) args.set(key, inlineValue);
    else if (next && !next.startsWith("--")) {
      args.set(key, next);
      i += 1;
    } else args.set(key, "true");
  }
}

const mode = args.get("mode") || "self-check";
const profile = loadProfile();
const fixtures = loadFixtures();

if (mode === "self-check") {
  validateProfile(profile);
  validateFixtures(fixtures);
  console.log(JSON.stringify({ ok: true, mode, profile, fixtures: fixtureSummary(fixtures) }, null, 2));
} else if (mode === "webapi") {
  await runWebApi(profile, fixtures);
} else if (mode === "sdk-js") {
  await runJsSdk(profile, fixtures);
} else {
  fail(`Unsupported mode: ${mode}`);
}

function loadProfile() {
  const fromFile = process.env.AICHAIN_PROFILE_FILE;
  if (fromFile) return JSON.parse(fs.readFileSync(path.resolve(fromFile), "utf8"));
  if (process.env.AICHAIN_PROFILE) return JSON.parse(process.env.AICHAIN_PROFILE);
  return {
    region: { code: process.env.AICHAIN_REGION || "cn", endpoint: process.env.AICHAIN_ENDPOINT || null },
    integration: mode === "sdk-js" ? "js_sdk" : "webapi",
    modelId: process.env.AICHAIN_MODEL_ID || "b16924f42ffe4fd895d4ba4778278bc3",
    capabilities: { asr: false, nlu: true, tts: false, image: false },
    asr: { language: null, duplex: null, vad: null },
    interruption: "none",
    audio: { input: null, output: null, notes: null },
    tts: { chunk: null },
    runtime: {
      appIdEnv: "AICHAIN_APP_ID",
      appKeyEnv: "AICHAIN_APP_KEY",
      endpointEnv: "AICHAIN_ENDPOINT",
      snEnv: "AICHAIN_SN",
    },
    android: {
      language: null,
      ui: null,
      minSdk: null,
      targetSdk: null,
      credentialStrategy: null,
      testStrategy: null,
    },
  };
}

function loadFixtures() {
  return {
    textZh: readText("assets/text/sample_zh.txt"),
    textEn: readText("assets/text/sample_en.txt"),
    zhAudio: readBytes("assets/audio/zh_16k_pcm.wav"),
    enAudio: readBytes("assets/audio/en_16k_pcm.wav"),
    image: readBytes("assets/image/apple.png"),
    imageQuestion: readText("assets/image/apple_question_zh.txt").trim(),
    imageExpected: readText("assets/image/apple_expected_zh.txt").trim(),
  };
}

function readText(relativePath) {
  return fs.readFileSync(path.join(root, relativePath), "utf8");
}

function readBytes(relativePath) {
  return fs.readFileSync(path.join(root, relativePath));
}

function validateProfile(value) {
  const missing = [];
  if (!value || typeof value !== "object") missing.push("profile");
  if (!value?.region?.code) missing.push("region.code");
  if (!value?.integration) missing.push("integration");
  if (!value?.modelId) missing.push("modelId");
  for (const key of ["asr", "nlu", "tts", "image"]) {
    if (typeof value?.capabilities?.[key] !== "boolean") missing.push(`capabilities.${key}`);
  }
  if (!value?.runtime?.appIdEnv) missing.push("runtime.appIdEnv");
  if (!value?.runtime?.appKeyEnv && !value?.runtime?.apiKeyEnv) missing.push("runtime.appKeyEnv");
  if (!value?.runtime?.endpointEnv) missing.push("runtime.endpointEnv");
  if (missing.length) fail(`Invalid AIChainProfile. Missing: ${missing.join(", ")}`);
}

function validateFixtures(value) {
  const checks = [
    ["textZh", value.textZh.length > 0],
    ["textEn", value.textEn.length > 0],
    ["zhAudio", isWav(value.zhAudio)],
    ["enAudio", isWav(value.enAudio)],
    ["image", value.image.subarray(0, 8).equals(Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))],
    ["imageQuestion", value.imageQuestion.length > 0],
    ["imageExpected", value.imageExpected.length > 0],
  ];
  const failed = checks.filter(([, ok]) => !ok).map(([name]) => name);
  if (failed.length) fail(`Invalid bundled fixtures: ${failed.join(", ")}`);
}

function isWav(buffer) {
  return buffer.subarray(0, 4).toString("ascii") === "RIFF" && buffer.subarray(8, 12).toString("ascii") === "WAVE";
}

function fixtureSummary(value) {
  return {
    textZhBytes: Buffer.byteLength(value.textZh),
    textEnBytes: Buffer.byteLength(value.textEn),
    zhAudioBytes: value.zhAudio.length,
    enAudioBytes: value.enAudio.length,
    imageBytes: value.image.length,
    imageQuestion: value.imageQuestion,
    imageExpected: value.imageExpected,
  };
}

function requireRuntime(profileValue) {
  const endpointEnv = profileValue.runtime.endpointEnv;
  const appIdEnv = profileValue.runtime.appIdEnv;
  const appKeyEnv = profileValue.runtime.appKeyEnv || profileValue.runtime.apiKeyEnv || "AICHAIN_APP_KEY";
  const fallbackAppKeyEnv = appKeyEnv === "AICHAIN_API_KEY" ? "AICHAIN_APP_KEY" : "AICHAIN_API_KEY";
  const endpoint = process.env[endpointEnv] || resolveEndpoint(profileValue.region?.code);
  const missing = [];
  if (!endpoint) missing.push(`${endpointEnv} or mapped AICHAIN_REGION`);
  if (!process.env[appIdEnv]) missing.push(appIdEnv);
  if (!process.env[appKeyEnv] && !process.env[fallbackAppKeyEnv]) missing.push(`${appKeyEnv} or ${fallbackAppKeyEnv}`);
  if (missing.length) fail(`Missing required environment variables: ${missing.join(", ")}`);
  return {
    endpoint,
    appId: process.env[appIdEnv],
    appKey: process.env[appKeyEnv] || process.env[fallbackAppKeyEnv],
    userId: process.env.AICHAIN_USER_ID,
    sn: process.env.AICHAIN_SN || process.env.AICHAIN_DEVICE_ID || "codex-smoke-001",
    scene: process.env.AICHAIN_SCENE || "main",
  };
}

function resolveEndpoint(regionCode) {
  const endpoints = {
    cn: "wss://aichain-sh.xfyun.cn",
    shanghai: "wss://aichain-sh.xfyun.cn",
    us: "wss://aichain-us.iflyoversea.com",
    rus: "wss://aichain-rus.iflyoversea.com",
  };
  return endpoints[String(regionCode || "cn").toLowerCase()] || null;
}

async function runWebApi(profileValue, fixtureValue) {
  validateProfile(profileValue);
  validateFixtures(fixtureValue);
  const runtime = requireRuntime(profileValue);
  if (!globalThis.WebSocket) fail("WebAPI mode requires Node.js with global WebSocket support.");

  const url = buildAiChainUrl(runtime);
  const transcript = [];
  let sid = null;
  let configed = false;
  const ws = new WebSocket(url);

  const timeout = Number(process.env.AICHAIN_TIMEOUT_MS || 30000);
  const done = new Promise((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error(`Timed out after ${timeout}ms`)), timeout);
    ws.addEventListener("open", () => {
      transcript.push("[open]");
    });
    ws.addEventListener("message", (event) => {
      const text = String(event.data);
      transcript.push(text);
      console.log(text);
      let data;
      try {
        data = JSON.parse(text);
      } catch {
        return;
      }

      if (data.type === "session.created") {
        sid = data.sid;
        ws.send(JSON.stringify({ type: "session.config", sid, config: buildSessionConfig(profileValue) }));
        return;
      }

      if (data.type === "session.configed") {
        configed = true;
        const cid = buildCid(runtime.appId);
        for (const message of buildConversationMessages(profileValue, fixtureValue, sid, cid)) {
          ws.send(JSON.stringify(message));
        }
        return;
      }

      if (data.type === "session.error") {
        clearTimeout(timer);
        ws.close();
        reject(new Error(`AIChain session.error: ${JSON.stringify(data.error || data)}`));
        return;
      }

      if (data.type === "event.cid_end") {
        clearTimeout(timer);
        ws.close();
        resolve();
      }
    });
    ws.addEventListener("error", () => {
      clearTimeout(timer);
      reject(new Error(`WebSocket error while connecting to ${redactUrl(url)}`));
    });
    ws.addEventListener("close", () => {
      clearTimeout(timer);
      if (!transcript.length) reject(new Error("Connection closed before any response"));
      else if (!sid) reject(new Error("Connection closed before session.created"));
      else if (!configed) reject(new Error("Connection closed before session.configed"));
    });
  });

  await done;
  validateServiceOutput(profileValue, transcript.join("\n"), fixtureValue);
}

function buildAiChainUrl(runtime) {
  const protocol = runtime.endpoint.toLowerCase().startsWith("ws://") ? "ws" : "wss";
  const host = runtime.endpoint.replace(/^wss?:\/\//i, "").split("/")[0];
  const curtime = Math.floor(Date.now() / 1000);
  const checksum = crypto.createHash("sha256").update(`${runtime.appKey}${curtime}`, "utf8").digest("hex");
  const params = new URLSearchParams({
    curtime: String(curtime),
    checksum,
    sn: runtime.sn,
    scene: runtime.scene,
  });
  return `${protocol}://${host}/v1/chat/${runtime.appId}?${params.toString()}`;
}

function redactUrl(url) {
  return url.replace(/checksum=[^&]+/g, "checksum=<redacted>");
}

function buildSessionConfig(profileValue) {
  const asrEnabled = Boolean(profileValue.capabilities.asr);
  const nluEnabled = Boolean(profileValue.capabilities.nlu);
  const ttsEnabled = Boolean(profileValue.capabilities.tts);
  return {
    mode: profileValue.asr?.duplex === "full" ? "full_duplex" : "half_duplex",
    simplifiedResponse: true,
    multiTurnEnabled: true,
    stt: asrEnabled
      ? {
          enable: true,
          language: profileValue.asr.language || "zh",
          audioConfig: buildAudioConfig(profileValue.audio?.input),
          vad: profileValue.asr.vad && profileValue.asr.vad !== "none"
            ? {
                enable: true,
                minSilenceDuration: profileValue.asr.vad === "acoustic_semantic" ? 300 : 600,
              }
            : { enable: false },
          turnDetection: profileValue.asr.vad === "acoustic_semantic"
            ? { enable: true, minEndpointingDelay: 250 }
            : { enable: false },
          interrupt: { enable: profileValue.interruption !== "none" },
        }
      : { enable: false },
    nlu: nluEnabled
      ? {
          enable: true,
          streamingFlushInterval: 300,
          tools: {
            model: {
              modelId: profileValue.modelId,
              maxTokens: 1000,
              temperature: 0.7,
              topP: 0.9,
            },
          },
        }
      : { enable: false },
    tts: ttsEnabled
      ? {
          enable: true,
          audioConfig: buildAudioConfig(profileValue.audio?.output),
          extraConfig: profileValue.tts?.chunk === null ? undefined : { chunk_tts: profileValue.tts?.chunk },
        }
      : { enable: false },
  };
}

function buildAudioConfig(codec) {
  if (!codec) return undefined;
  return {
    audioEncoding: codec === "pcm" ? "raw" : codec,
    format: "plain",
    sampleRate: 16000,
    bitDepth: 16,
    channels: 1,
  };
}

function buildConversationMessages(profileValue, fixtureValue, sid, cid) {
  const items = [];
  if (profileValue.capabilities.nlu || profileValue.capabilities.tts) {
    items.push({ type: "text", data: fixtureValue.textZh });
  }
  if (profileValue.capabilities.asr) {
    const audio = selectAudio(profileValue, fixtureValue);
    items.push({ type: "audio", data: audio.toString("base64") });
  }
  if (profileValue.capabilities.image) {
    items.push({ type: "text", data: fixtureValue.imageQuestion });
    items.push({ type: "image", data: fixtureValue.image.toString("base64"), extend: { mimeType: "image/png" } });
  }
  return [{ type: "conversation.user.append", sid, cid, items, endFlag: true }];
}

function buildCid(appId) {
  const now = new Date();
  const ts = [
    now.getUTCFullYear(),
    String(now.getUTCMonth() + 1).padStart(2, "0"),
    String(now.getUTCDate()).padStart(2, "0"),
    String(now.getUTCHours()).padStart(2, "0"),
    String(now.getUTCMinutes()).padStart(2, "0"),
    String(now.getUTCSeconds()).padStart(2, "0"),
  ].join("");
  return `${appId}@${ts}-${crypto.randomBytes(3).toString("hex")}`;
}

async function runJsSdk(profileValue, fixtureValue) {
  validateProfile(profileValue);
  validateFixtures(fixtureValue);
  const runtime = requireRuntime(profileValue);
  const packageName = process.env.AICHAIN_JS_SDK_PACKAGE;
  if (!packageName) fail("Missing AICHAIN_JS_SDK_PACKAGE for generic JS SDK mode.");

  const sdk = await import(packageName);
  const Client = sdk.default || sdk.AIChainClient || sdk.Client;
  if (!Client) fail(`Cannot find default export, AIChainClient, or Client in ${packageName}`);

  const client = new Client({
    endpoint: runtime.endpoint,
    appId: runtime.appId,
    appKey: runtime.appKey,
    userId: runtime.userId,
    sn: runtime.sn,
    modelId: profileValue.modelId,
    profile: profileValue,
  });

  const results = [];
  if (profileValue.capabilities.nlu || profileValue.capabilities.tts) {
    results.push(await callFirst(client, ["sendText", "chat"], fixtureValue.textZh, { text: fixtureValue.textZh, profile: profileValue }));
  }
  if (profileValue.capabilities.asr) {
    results.push(await callFirst(client, ["sendAudio", "recognize"], selectAudio(profileValue, fixtureValue), { audio: selectAudio(profileValue, fixtureValue), profile: profileValue }));
  }
  if (profileValue.capabilities.image) {
    results.push(await callFirst(client, ["sendImage", "chat"], fixtureValue.image, { image: fixtureValue.image, prompt: fixtureValue.imageQuestion, mimeType: "image/png", profile: profileValue }));
  }

  const output = JSON.stringify(results, null, 2);
  console.log(output);
  validateServiceOutput(profileValue, output, fixtureValue);
}

async function callFirst(client, names, positional, objectArg) {
  for (const name of names) {
    if (typeof client[name] === "function") {
      return client[name].length <= 1 ? client[name](objectArg) : client[name](positional, objectArg);
    }
  }
  fail(`SDK client must expose one of: ${names.join(", ")}`);
}

function selectAudio(profileValue, fixtureValue) {
  if (profileValue.asr?.language === "en") return fixtureValue.enAudio;
  return fixtureValue.zhAudio;
}

function validateServiceOutput(profileValue, output, fixtureValue) {
  const lower = output.toLowerCase();
  const failures = [];
  if (profileValue.capabilities.asr) {
    const expected = profileValue.asr?.language === "en" ? ["hello", "weather"] : ["北京", "天气"];
    for (const term of expected) {
      if (!lower.includes(term.toLowerCase())) failures.push(`ASR output missing ${term}`);
    }
  }
  if (profileValue.capabilities.nlu && output.trim().length === 0) failures.push("NLU output is empty");
  if (profileValue.capabilities.tts && !/audio|tts|pcm|opus|wav|base64/i.test(output)) failures.push("TTS output did not mention or contain audio");
  if (profileValue.capabilities.image && !lower.includes(fixtureValue.imageExpected.toLowerCase())) {
    failures.push(`Image output missing ${fixtureValue.imageExpected}`);
  }
  if (failures.length) fail(`AIChain E2E validation failed:\n- ${failures.join("\n- ")}`);
}

function fail(message) {
  console.error(message);
  process.exit(1);
}
